### PR TITLE
Fix Windows shared lib build

### PR DESCRIPTION
**Pull request by ocefpaf**
Fixes #2

### CODE REVIEW

**Comment by ocefpaf** [comment]
@gillins can you take a look at the `bld.bat`? I am not sure if the copying/moving of the headers is OK.
@ccordoba12 did you write the original? Can you tell me why that is needed?

PS: I was unsuccessful to build pillow and matplotlib with this locally. Not sure if it is my Windows VM or a problem with the package. I get several missing unresolved `_png<something>` symbols. (I cannot copy and paste from my Windows VM :unamused: )

**Comment by ccordoba12** [comment]
@msarahan is in charge of the Windows build :-)

**Comment by msarahan** [comment]
I think the file copying looks fine.  I don't know exactly how programs typically do their includes here, though.  Basically, what you have says includes will be:

`#include <"ft2build.h">`

if includes are rather:

`#include <"freetype2/ft2build.h">`

then you need to keep the copying/moving more as it was: creating the freetype2 folder in %LIBRARY_INC%\

**Comment by ocefpaf** [comment]
Thanks @msarahan. I tried to mimic the final structure I saw in the default channel version. The configuration we have here now is OK for `pillow` and `matplotlib`. OK in the sense that they can find the headers, but I am having other issues (https://github.com/conda-forge/freetype-feedstock/pull/3#issuecomment-218223414).

**Comment by msarahan** [comment]
Points to PNG linker issues, obviously. Might be expecting a static lib when you are providing a dynamic one, or vice versa.  Might need extra flags or preprocessor defines here, or on the other packages, or maybe both.  I'll try to take a look soon, but I'm pretty tired up. :(

**Comment by ocefpaf** [comment]
> I'll try to take a look soon, but I'm pretty tired up. :(

I know! conda-forge alone is tiresome and you still have a day job to go back to :wink: 

Don't worry about it. I will take a second look. I am trying to learn more about Windows builds, cmake, and all that jazz.

**Comment by gillins** [comment]
Problem was indeed related to static/dynamic. Freetype was being built statically and the linker was trying to grab all the png symbols from it. I fixed it for me by specifying `-D BUILD_SHARED_LIBS:BOOL=true` in bld.bat and this patch:
[winshared.patch.txt](https://github.com/conda-forge/freetype-feedstock/files/260398/winshared.patch.txt)

Building with VS rather than mingw seems to be getting harder by the day!

I'm still getting a matplotlib error where it is trying to link against `png.lib` rather than `libpng` for some reason - I'll look into this next.

**Comment by gillins** [comment]
I think this change to matplotlib's `setupext.py` should work:

```
         pkg_config.setup_extension(
-            ext, 'libpng', default_libraries=['png', 'z'],
+            ext, 'libpng', default_libraries=['libpng', 'z'],
```

I'm not sure how you want to manage these changes....

**Comment by ocefpaf** [comment]
> Problem was indeed related to static/dynamic. Freetype was being built statically and the linker was trying to grab all the png symbols from it. I fixed it for me by specifying `-D BUILD_SHARED_LIBS:BOOL=true` in bld.bat and this patch:
> [winshared.patch.txt](https://github.com/conda-forge/freetype-feedstock/files/260398/winshared.patch.txt)

Thanks for the patch @gillins! Putting it here now.

> Building with VS rather than mingw seems to be getting harder by the day!

:smile: 

> I'm still getting a matplotlib error where it is trying to link against `png.lib` rather than `libpng` for some reason - I'll look into this next.

Looks like you already nailed that too. Thanks!

File: recipe/bld.bat [hunk 1]
```diff
@@ -23,5 +24,5 @@ nmake install
 if errorlevel 1 exit 1
 
 :: Move everything 1-level down.
-xcopy %LIBRARY_INC%\freetype2\freetype\*.* %LIBRARY_INC%\freetype2\ /s /e || exit 1
-rmdir %LIBRARY_INC%\freetype2\freetype /s /q
+move %LIBRARY_INC%\freetype2\freetype %LIBRARY_INC% || exit 1
+move %LIBRARY_INC%\freetype2\ft2build.h %LIBRARY_INC% || exit 1
```
Comment (pelson) [inline recipe/bld.bat:28]: What is driving this behaviour in the first place? If freetype does it one particular way, should we honour that? After all, downstream builders need to deal with the weird folder choice for non conda builds, no?

Comment (pelson) [inline recipe/bld.bat:28]: P.S. I'm just raising the question, not intending to block anything here. I'm completely OK with doing it, just want to make sure we are not being wagged by the tail (context for those who don't have English as their first language: the expression "tail wagging the dog")

Comment (ocefpaf) [inline recipe/bld.bat:28]: > What is driving this behaviour in the first place?

That is a good question. I did that only to mimic the default channel version of the package. I have no idea if that is the best approach.

**Comment by ocefpaf** [comment]
I am tempted to merge this and check how `pillow` and `matplotlib` will behave. If the directory for headers is a problem we can re-visit the issue.

File: recipe/bld.bat (hunk 1 above)
Comment (jakirkham) [inline recipe/bld.bat:28]: From my understanding this is the correct thing to do. When it comes to Windows `conda` does some things different, but I get the impression they are doing it so there is a standard to follow as opposed to [several](https://xkcd.com/927/). 😉

**Pull request merged**